### PR TITLE
Increase voting cap from 30 to Unlimited

### PR DESCRIPTION
**Pull request by jafri**

### CODE REVIEW

**Comment by arhag** [comment]
A strict cap on the number of BPs an account can vote on is necessary unless we redesign the per-vote BP reward mechanism, because otherwise it opens up an attack vector that allows malicious stakeholders with sufficiently high stake to consume an extremely disproportionate fraction of the per-vote BP rewards.

The basic mechanism of the attack is for a colluding group of large stakeholders (collectively with sufficient stake to cause an account they vote for to rise from nothing to a ranking high enough to collect per-vote BP rewards) to create a large number of accounts and vote for all of them with their collective stake. 

Let's say the natural rank threshold to get paid some amount of BP rewards happens to be (and stays even after the attack) at 100. If the colluding group has enough stake to get an account they vote for to, for example, rank 26, then they simply need to create at least 75 accounts and vote for them all. Those 75 accounts will collect a very large share of the per-vote BP rewards with the remaining going to the top 25. But with the limit of 30 votes per account, they could only vote with their full collective stake for just 30 of those 75 accounts. They would still be paid per-vote BP rewards for 30 of 100 paid accounts, which is still disproportionate, but the ratio is not as extreme as what it could potentially be without any limits.

The fundamental problem here is that approval-at-large voting, while a decent mechanism for determining the top 21 set that become active block producers (and get paid per-block rewards), is not the best fit for the per-vote BP rewards mechanism. In fact, because of the limit of 30 candidates per voter (put in place as a compromise to protect attacks against the per-vote BP rewards system), our system is not actually an approval-at-large voting system for electing the 21 active block producers. It is something in between that and a plurality-at-large voting system (also known as block voting). Perhaps it might be necessary to use different voting mechanisms for determining per-vote BP rewards and determining the 21 active block producers to avoid the compromises.

A voting mechanism that allows voters to distribute their stake-proportional votes among any number of candidates (with the sum never exceeding their total allowed votes) may be a better system for determining per-vote BP rewards; such a system is called cumulative voting (with the points method). However, that system has disadvantages compared to approval-at-large voting when it comes to deciding the top 21 active block producers: it incentivizes plumping, or just focusing all the votes to just one candidate (most likely oneself) to maximize their likelihood of becoming a BP and/or winning more per-vote BP rewards. 

A possible adjustment to the cumulative voting described above could be to also put a max limit on the vote/points per any one candidate to some fraction of the voter's total allowed votes. So, for example, the voter may get points equal to 30 times their stake and they are allowed to distribute that over as many candidates as they want in whichever way they wish with only the limitation that no candidate can receive more points from that voter than the amount of stake that they have.  This may or may not be an appropriate compromise between the needs of per-vote BP rewards and the needs of selecting the 21 active block producers.

Then there may be other voting systems (or combinations of them for the different purposes) that provide other desirable properties, such as proportional representation. For example, perhaps a proportional approval voting system would be ideal for selecting the 21 active block producers. But even putting aside the possibility that another mechanism may be necessary for determining per-vote BP rewards, that approach may come with other challenges such as incurring a computational cost that is too severe to pull off in a live voting system running entirely in blockchain smart contracts. 

In conclusion, a more fundamental change to the design of the voting mechanism is required to get to an overall much better solution where hopefully we don't sacrifice the security we expect for determining the 21 active block producers, disallow the kinds of abuses described above that allow colluding stakeholders to get more compensation via BP rewards than their "fair" share (however we define that word), and maybe get other desirable properties like proportional representation and better resistance to tactical voting. But don't expect that anytime soon (also it is not clear whether the combinations of all the desirable properties we care about are even possible). In the mean time, we are sticking with our current system with the 30 candidate voting cap enforced.
